src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,443 @@
+            window_size: The size of the window to calculate the average over.
+            weights: The weights to use for the average calculation.
+
+        Returns:
+            The averaged window.
+        """
+        num_windows = self._num_windows(window, window_size)
+
+        if len(window) - num_windows * self._distance >= window_size:
+            num_windows += 1
+
+        res = np.empty(window_size)
+
+        for i in range(window_size):
+            assert num_windows * (self._distance) - len(window) <= self._distance
+            summe = 0
+            for j in range(num_windows):
+                if weights is None:
+                    summe += window[i + (j * (self._distance))]
+                else:
+                    summe += weights[j] * window[i + (j * (self._distance))]
+
+            if weights is None:
+                res[i] = summe / num_windows
+            else:
+                res[i] = summe / np.sum(weights)
+
+        return res
+
+    def _get_buffer_bounds(
+        self, start: datetime, end: datetime
+    ) -> Tuple[int, int, int]:
+        """
+        Get the bounds of the ringbuffer used for further operations.
+
+        This method uses the given start and end timestamps to calculate
+        the part of the ringbuffer that can be used for further operations,
+        like average or min/max.
+
+        Here we cut out the oldest window or the window that is currently overwritten
+        in the MovingWindow such that it is not considered in any further operation.
+
+        Args:
+            start: The start timestamp of the window.
+            end: The end timestamp of the window.
+
+        Returns:
+            The current positions of the ringbuffer.
+
+        Raises:
+            ValueError: If the start timestamp is after the end timestamp.
+        """
+        if end <= start:
+            raise ValueError("Start timestamp must be before end timestamp")
+
+        # The number of usable windows can change, when the current position
+        # of the ringbuffer is inside one of the windows inside the MovingWindow.
+        # Since this is possible, we assume that one window is always not used
+        # for the average calculation.
+        # We are ignoring either the window that is currently overwritten
+        # if the current position is inside that window or the window
+        # that would be overwritten next.
+
+        # Move the window to its first appereance in the MovingWindow
+        # relative to the oldest sample stored in the MovingWindow.
+        # In other words the oldest stored sample is considered to have index 0.
+        # Note that the returned value is a index not a timestamp
+        rel_start_sample = self._timestamp_to_rel_index(start) % self._distance
+        rel_end_sample = self._timestamp_to_rel_index(end) % self._distance
+        # fix the position if modulo period the end_sample
+        # is smaler then the start sample position
```

Comment:
what does "if modulo period the end_sample" mean?

``` suggestion
        # fix the position if modulo period the end_sample
        # is smaller then the start sample position

Reply:
it means the `rel_next_position % self.period`